File: packages/gbnf-experiment/data/20260909T133702Z_8305aa82/ported_implementation/gbnf/utils/errors/input_parse_error.py
from ...grammar_graph.types import ValidInput
from .build_error_position import build_error_position
from .get_input_as_string import get_input_as_string

INPUT_PARSER_ERROR_HEADER_MESSAGE = "Failed to parse input string:"


class InputParseError(Exception):
    def __init__(
        self,
        most_recent_input: ValidInput,
        pos: int,
        previous_input: ValidInput = "",
    ):
        self._most_recent_input = most_recent_input
        self._pos = pos
        self._previous_input = previous_input
        previous_as_string = get_input_as_string(previous_input)
        message = "\n".join(
            [
                INPUT_PARSER_ERROR_HEADER_MESSAGE,
                "",
                *build_error_position(
                    f"{previous_as_string}{get_input_as_string(most_recent_input)}",
                    pos + len(previous_as_string),
                ),
            ]
        )
        super().__init__(message)
        self.name = "InputParseError"
        self.message = message

    def __str__(self) -> str:
        return self.message

    @property
    def src(self) -> str:
        return (
            f"{get_input_as_string(self._previous_input)}"
            f"{get_input_as_string(self._most_recent_input)}"
        )

    @property
    def error_for_most_recent_input(self) -> str:
        return "\n".join(
            [
                INPUT_PARSER_ERROR_HEADER_MESSAGE,
                "",
                *build_error_position(
                    get_input_as_string(self._most_recent_input), self._pos
                ),
            ]
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, InputParseError):
            return NotImplemented
        return self.message == other.message and self.src == other.src

    def __hash__(self) -> int:
        return hash(self.message)

File: packages/gbnf-experiment/data/20260909T181751Z_00b43b48/ported_implementation/src/utils/errors/input-parse-error.ts
import { buildErrorPosition } from './build-error-position.js';
import type { ValidInput } from './errors-types.js';
import { getInputAsString } from './get-input-as-string.js';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

export class InputParseError extends Error {
  mostRecentInput: ValidInput;
  pos: number;
  previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    super([
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(
        `${getInputAsString(previousInput)}${getInputAsString(mostRecentInput)}`,
        pos + getInputAsString(previousInput).length,
      ),
    ].join('\n'));
    this.name = 'InputParseError';
    this.mostRecentInput = mostRecentInput;
    this.pos = pos;
    this.previousInput = previousInput;
    Object.setPrototypeOf(this, InputParseError.prototype);
  }

  get src(): string {
    return `${getInputAsString(this.previousInput)}${getInputAsString(this.mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return [
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(getInputAsString(this.mostRecentInput), this.pos),
    ].join('\n');
  }
}
